python/src/iceberg/table/metadata.py

```diff
@@ -0,0 +1,259 @@
+    the expected UUID after refreshing metadata."""
+
+    format_version: Literal[1] = Field(alias="format-version")
+    """An integer version number for the format. Currently, this can be 1 or 2
+    based on the spec. Implementations must throw an exception if a table’s
+    version is higher than the supported version."""
+
+    schema_: Schema = Field(alias="schema")
+    """The table’s current schema. (Deprecated: use schemas and
+    current-schema-id instead)"""
+
+    partition_spec: List[Dict[str, Any]] = Field(alias="partition-spec")
+    """The table’s current partition spec, stored as only fields.
+    Note that this is used by writers to partition data, but is
+    not used when reading because reads use the specs stored in
+    manifest files. (Deprecated: use partition-specs and default-spec-id
+    instead)"""
+
+
+class TableMetadataV2(TableMetadataCommonFields, IcebergBaseModel):
+    @root_validator(skip_on_failure=True)
+    def check_if_schema_is_found(cls, data: Dict[str, Any]):
+        current_schema_id = data["current_schema_id"]
+
+        for schema in data["schemas"]:
+            if schema.schema_id == current_schema_id:
+                return data
+
+        raise ValueError(f"current-schema-id {current_schema_id} can't be found in the schemas")
+
+    @root_validator
+    def check_partition_spec(cls, data: Dict[str, Any]):
+        default_spec_id = data["default_spec_id"]
+
+        for spec in data["partition_specs"]:
+            if spec["spec-id"] == default_spec_id:
+                return data
+
+        raise ValueError(f"default-spec-id {default_spec_id} can't be found")
+
+    @root_validator(skip_on_failure=True)
+    def check_sort_order(cls, data: Dict[str, Any]):
+        default_sort_order_id = data["default_sort_order_id"]
+
+        # 0 == unsorted
+        if default_sort_order_id != 0:
+            for sort in data["sort_orders"]:
+                if sort["order-id"] == default_sort_order_id:
+                    return data
+
+            raise ValueError(f"default-sort-order-id {default_sort_order_id} can't be found")
+        return data
+
+    format_version: Literal[2] = Field(alias="format-version")
+    """An integer version number for the format. Currently, this can be 1 or 2
+    based on the spec. Implementations must throw an exception if a table’s
+    version is higher than the supported version."""
+
+    table_uuid: UUID = Field(alias="table-uuid")
+    """A UUID that identifies the table, generated when the table is created.
+    Implementations must throw an exception if a table’s UUID does not match
+    the expected UUID after refreshing metadata."""
+
+    last_sequence_number: int = Field(alias="last-sequence-number", default=_INITIAL_SEQUENCE_NUMBER)
+    """The table’s highest assigned sequence number, a monotonically
+    increasing long that tracks the order of snapshots in a table."""
+
+    refs: Dict[str, SnapshotRef] = Field(default_factory=dict)
```

Comment:
This can be in common metadata. There's no reason why v1 tables can't also have refs. Adding refs is a forward-compatible change.

Reply:
Fair point, I've just moved them.